src/main/scala/nl/biopet/tools/downsampleregions/DownsampleRegions.scala

```diff
@@ -0,0 +1,133 @@
+
+    val paired = {
+      val it = bamReader.iterator()
+      val firstRecord = it.next()
+      it.close()
+      firstRecord.getReadPairedFlag
+    }
+    require(paired == cmdArgs.inputR2.isDefined, "Bam contains paired reads but input R2 is not defined")
+    require(paired == cmdArgs.outputR2A.isDefined, "Bam contains paired reads but output A R2 is not defined")
+    require(paired == cmdArgs.outputR2B.isDefined, "Bam contains paired reads but output B R2 is not defined")
+
+    var totalReads = 0L
+    val removeIdsA: mutable.Set[String] = mutable.Set()
+    val removeIdsB: mutable.Set[String] = mutable.Set()
+    for (region <- regions.allRecords) {
+      val fraction = region.score.get + (cmdArgs.deviation * (Random.nextDouble() - 0.5))
+      val bamIt = bamReader.query(region.chr, region.start, region.end, false)
+      for (samRecord <- bamIt) {
+        def removeRead() = {
+          totalReads += 1
+          val remove = Random.nextDouble() <= fraction.abs
+          if (remove) {
+            if (region.score.get >= 0.0 ) removeIdsA += samRecord.getReadName
+            else removeIdsB += samRecord.getReadName
+          }
+        }
+        if (paired) {
+          if (!samRecord.isSecondaryOrSupplementary && samRecord.getFirstOfPairFlag) removeRead()
+        } else removeRead()
+      }
+      bamIt.close()
+    }
+    bamReader.close()
+
+    logger.info(s"Found $totalReads reads")
+    logger.info(s"Will remove ${removeIdsA.size} reads in A")
+    logger.info(s"Will remove ${removeIdsB.size} reads in B")
+
+    val readerR1 = new FastqReader(cmdArgs.inputR1)
+    val writerR1A = new AsyncFastqWriter(new BasicFastqWriter(cmdArgs.outputR1A), 1000)
+    val writerR1B = new AsyncFastqWriter(new BasicFastqWriter(cmdArgs.outputR1B), 1000)
+
+    for (record <- readerR1.iterator()) {
+      if (!removeIdsA.contains(record.getReadName.split(" ").head.stripSuffix("/1")))
+        writerR1A.write(record)
+      if (!removeIdsB.contains(record.getReadName.split(" ").head.stripSuffix("/1")))
+        writerR1B.write(record)
+    }
+    readerR1.close()
+    writerR1A.close()
+    writerR1B.close()
+
+    if (paired) {
+      val readerR2 = new FastqReader(cmdArgs.inputR2.get)
+      val writerR2A = new AsyncFastqWriter(new BasicFastqWriter(cmdArgs.outputR2A.get), 1000)
+      val writerR2B = new AsyncFastqWriter(new BasicFastqWriter(cmdArgs.outputR2B.get), 1000)
+
+      for (record <- readerR2.iterator()) {
+        if (!removeIdsA.contains(record.getReadName.split(" ").head.stripSuffix("/2")))
+          writerR2A.write(record)
+        if (!removeIdsB.contains(record.getReadName.split(" ").head.stripSuffix("/2")))
+          writerR2B.write(record)
+      }
+
+      readerR2.close()
+      writerR2A.close()
+      writerR2B.close()
+    }
+
+    logger.info("Done")
+  }
+
+  def argsParser = new ArgsParser(toolName)
+
+  def emptyArgs = Args()
+
+  def descriptionText: String =
+    """
```

Comment:
Better to include the documentation at the top for readability.

Reply:
Don't think I agree completely with this. Once someone is ready to look at the code they want to see the working code. This method are very good to have it there but there are just for generation of the code. What does miss here is maybe some scaladocs ;)